source/elements/oneDAL/source/algorithms/nearest_neighbors/knn_classification.rst

```diff
@@ -1,3 +1,198 @@
+:math:`X` and :math:`X'` at the inference stage.
+
+
+.. _t_math_brute_force:
+
+Training method: *brute-force*
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The training operation produces the model that simply stores all the
+feature vectors from the initial training set :math:`X`.
+
+
+.. _t_math_kd_tree:
+
+Training method: *k-d tree*
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The training operation builds a :math:`k`-:math:`d` tree (for more details, see
+:ref:`kd_tree`) that partitions the training set :math:`X`.
+
+
+.. _i_math:
+
+Inference
+---------
+Given the set :math:`X' = \{ x_1', \ldots, x_m' \}` of :math:`p`-dimensional
+feature vectors, the user defined number of nearest neighbors :math:`k` and the
+model produced by the training operation. The inference operation predicts the
+label for each :math:`x_i'`, :math:`1 \leq i \leq n`, by performing the
+following steps:
+
+#. Identify the set :math:`N(x_i') \subseteq X` of the :math:`k` feature vectors
+   in the training set that are nearest to :math:`x_i'` with respect to the
+   Euclidean distance.
+
+#. Estimate the conditional probability for the :math:`l`-th class as the
+   fraction of vectors in :math:`N(x_i')` whose labels :math:`y_i` are equal to
+   :math:`l`:
+
+   .. math::
+      :label: p_predict
+
+      P(y_i' = l \; | \; x_i') = \frac{1}{| N(x_i') |}
+      \Big| \big\{ x_r \in N(x_i') : y_r = l \big\} \Big|,
+      \quad 1 \leq i \leq m, \; 0 \leq l < c.
+
+#. Use the class with the largest probability as prediction for the feature
+   vector :math:`x_i'`:
+
+   .. math::
+      :label: y_predict
+
+      y_i' = \mathrm{arg}\max_{0 \leq l < c} P(y_i' = l \; | \; x_i'),
+      \quad 1 \leq i \leq m.
+
+
+.. _i_math_brute_force:
+
+Inference method: *brute-force*
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The inference operation determines the set :math:`N(x_j')` of the nearest
+feature vectors by iterating over all the pairs :math:`(x_j', x_i)` in the
+implementation defined order, :math:`1 \leq i \leq n`, :math:`1 \leq j \leq m`.
+The final prediction is computed according to the equations :eq:`p_predict` and
+:eq:`y_predict`.
+
+
+.. _i_math_kd_tree:
+
+Inference method: *k-d tree*
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+
+
+---------------------
+Programming Interface
+---------------------
+
+Computational methods
+---------------------
+.. namespace:: onedal::classification::knn::method
+
+::
+
+   namespace method {
+      struct bruteforce {};
+      struct kd_tree {};
+      using by_default = brutforce;
+   } // namespace method
+
+
+.. type:: bruteforce
+
+   Tag-type that denotes `brute-force <t_math_brute_force_>`_ computational
+   method.
+
+
+.. type:: kd_tree
+
+   Tag-type that denotes `k-d tree <t_math_kd_tree>`_ computational method.
+
+
+.. type:: by_default = bruteforce
+
+   Alias tag-type for `brute-force <t_math_brute_force_>`_ computational method.
+
+
+Descriptor
+----------
+.. onedal_class:: classification::knn::desc
```

Comment:
> what would happen in the corner case with the algorithm, if number of classes is 1

Nothing special, the algorithm is expected to work, but always predict the one class, so this classifier is absolutely useless as it does not perform its main function.

I can image the only one more or less meaningful use case when the user wants to get the list of neighbors and does not care about labels. But for that case I'd prefer to have the different entry point.

Reply:
> Also, why the default value of neighbor_count is 5

I considered `1` (we have it in old APIs) and `5` (it's used in scikit-learn). I'd stick to sikit-learn, because `1` works only in the very specific problems. To me, `5` is better candidate for default value as it's expected to work out-of-the-box on the larger set of problems than `1`.

Reply:
The same question for `class_count`, there is no justification for default value. @andreynv, would you recommend to remove default parameters, and always pass `class_count`, `neighbor_count` to constructor?

Reply:
If there is no intuition on the values that might be used quite often, it would be better to remove the defaults as in 99% the apps would have their specific values for classes&neighbors